pkg/bot/lark.go

```diff
@@ -0,0 +1,137 @@
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package bot
+
+import (
+	"fmt"
+	"github.com/infracloudio/botkube/pkg/config"
+	"github.com/infracloudio/botkube/pkg/execute"
+	"github.com/infracloudio/botkube/pkg/log"
+	"github.com/infracloudio/botkube/pkg/utils"
+	"github.com/larksuite/oapi-sdk-go/core"
+	"github.com/larksuite/oapi-sdk-go/core/tools"
+	"github.com/larksuite/oapi-sdk-go/event"
+	eventhttpserver "github.com/larksuite/oapi-sdk-go/event/http/native"
+	"net/http"
+	"strings"
+)
+
+// LarkBot listens for user's message, execute commands and sends back the response
+type LarkBot struct {
+	AllowKubectl     bool
+	RestrictAccess   bool
+	ClusterName      string
+	DefaultNamespace string
+	Port             int
+	MessagePath      string
+	LarkClient       *utils.LarkClient
+}
+
+// NewLarkBot returns new Bot object
+func NewLarkBot(c *config.Config) Bot {
+	larkConf := c.Communications.Lark
+	appSettings := core.NewInternalAppSettings(core.SetAppCredentials(larkConf.AppID, larkConf.AppSecret),
+		core.SetAppEventKey(larkConf.VerificationToken, larkConf.EncryptKey))
+	conf := core.NewConfig(core.Domain(larkConf.Endpoint), appSettings, core.SetLoggerLevel(core.LoggerLevelError))
+	return &LarkBot{
+		AllowKubectl:     c.Settings.Kubectl.Enabled,
+		RestrictAccess:   c.Settings.Kubectl.RestrictAccess,
+		ClusterName:      c.Settings.ClusterName,
+		DefaultNamespace: c.Settings.Kubectl.DefaultNamespace,
+		Port:             c.Communications.Lark.Port,
+		MessagePath:      c.Communications.Lark.MessagePath,
+		LarkClient:       utils.NewLarkClient(conf),
+	}
+}
+
+// Execute execute commands sent by users
+func (l *LarkBot) Execute(e map[string]interface{}) {
+	event := e["event"].(map[string]interface{})
+
+	chatType := event["chat_type"].(string)
+	text := event["text_without_at_bot"].(string)
+
+	executor := execute.NewDefaultExecutor(text, l.AllowKubectl, l.RestrictAccess, l.DefaultNamespace,
+		l.ClusterName, config.LarkBot, "", true)
+	response := executor.Execute()
+
+	if chatType == "group" {
+		l.LarkClient.SendTextMessage("chat_id", event["open_chat_id"].(string), response)
+	} else {
+		l.LarkClient.SendTextMessage("open_id", event["open_id"].(string), response)
+	}
+}
+
+// Start starts the lark server and listens for lark messages
+func (l *LarkBot) Start() {
+	// See: https://open.larksuite.com/document/ukTMukTMukTM/ukjNxYjL5YTM24SO2EjN
+	// message
+	larkConf := l.LarkClient.Conf
+	event.SetTypeCallback(larkConf, "message", func(ctx *core.Context, e map[string]interface{}) error {
+		log.Infof(ctx.GetRequestID())
+		log.Infof(tools.Prettify(e))
+		go l.Execute(e)
+		return nil
+	})
+
+	// add_bot
+	event.SetTypeCallback(larkConf, "add_bot", func(ctx *core.Context, e map[string]interface{}) error {
+		log.Infof(ctx.GetRequestID())
+		log.Infof(tools.Prettify(e))
+		go l.SayHello(e)
```

Comment:
Why do we need to SayHello each time? What is the significance of this?

Reply:
I see some event callbacks. When these events occur, the Lark Bot needs to respond. Such as when a new member joins the group or when a member creates a session with the Lark Bot for the first time

Reply:
@Lion916 do we always need to respond to such events? Can't we ignore those? Because those are not relevant for us.